examples/parser/Parser/CSV.roc

```diff
@@ -0,0 +1,196 @@
+field : Parser RawStr a -> Parser CSVRecord a
+field = \fieldParser ->
+    buildPrimitiveParser \fieldsList ->
+        when List.get fieldsList 0 is
+            Err OutOfBounds ->
+                Err (ParsingFailure "expected another CSV field but there are no more fields in this record")
+
+            Ok rawStr ->
+                when Parser.Str.parseRawStr fieldParser rawStr is
+                    Ok val ->
+                        Ok { val: val, input: List.dropFirst fieldsList }
+
+                    Err (ParsingFailure reason) ->
+                        fieldStr = rawStr |> strFromRaw
+
+                        Err (ParsingFailure "Field `\(fieldStr)` could not be parsed. \(reason)")
+
+                    Err (ParsingIncomplete reason) ->
+                        reasonStr = strFromRaw reason
+                        fieldsStr = fieldsList |> List.map strFromRaw |> Str.joinWith ", "
+
+                        Err (ParsingFailure "The field parser was unable to read the whole field: `\(reasonStr)` while parsing the first field of leftover \(fieldsStr))")
+
+## Parser for a field containing a UTF8-encoded string
+string : Parser CSVField Str
+string = Parser.Str.anyString
+
+## Parse a natural number from a CSV field
+nat : Parser CSVField Nat
+nat =
+    string
+    |> map
+        (\val ->
+            when Str.toNat val is
+                Ok num ->
+                    Ok num
+
+                Err _ ->
+                    Err "\(val) is not a Nat."
+        )
+    |> flatten
+
+## Parse a 64-bit float from a CSV field
+f64 : Parser CSVField F64
+f64 =
+    string
+    |> map
+        (\val ->
+            when Str.toF64 val is
+                Ok num ->
+                    Ok num
+
+                Err _ ->
+                    Err "\(val) is not a F64."
+        )
+    |> flatten
+
+## Attempts to parse a Str into the internal `CSV` datastructure (A list of lists of bytestring-fields).
+parseStrToCSV : Str -> Result CSV [ParsingFailure Str, ParsingIncomplete RawStr]
+parseStrToCSV = \input ->
+    parse file (Str.toUtf8 input) (\leftover -> leftover == [])
+
+## Attempts to parse a Str into the internal `CSVRecord` datastructure (A list of bytestring-fields).
+parseStrToCSVRecord : Str -> Result CSVRecord [ParsingFailure Str, ParsingIncomplete RawStr]
+parseStrToCSVRecord = \input ->
+    parse csvRecord (Str.toUtf8 input) (\leftover -> leftover == [])
+
+# The following are parsers to turn strings into CSV structures
+file : Parser RawStr CSV
+file = sepBy csvRecord endOfLine
+
+csvRecord : Parser RawStr CSVRecord
+csvRecord = sepBy1 csvField comma
+
+csvField : Parser RawStr CSVField
+csvField = alt escapedCsvField nonescapedCsvField
+
+escapedCsvField : Parser RawStr CSVField
+escapedCsvField = between escapedContents dquote dquote
+escapedContents = many
+    (
+        oneOf [
+            twodquotes |> map (\_ -> 34), # An escaped double quote
```

Comment:
what we've done in the json parser is defined

```
asciiByte = \b -> Num.toU8 b
```

so that then you can do
```suggestion
            twodquotes |> map (\_ -> asciiByte '"'),
```
here, and similar for the comma, cr, lf, etc. values defined below.

In the future this should be replaced with the integer-suffixed character literals, but we don't have those yet.